tests/0102-static_group_rebalance.c

```diff
@@ -0,0 +1,297 @@
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "test.h"
+
+
+/**
+ * @name KafkaConsumer static membership tests
+ *
+ * Runs two consumers subscribing to a topic simulating various
+ * rebalance scenarios.
+ */
+
+#define _CONSUMER_CNT 2
+
+typedef struct _consumer_s {
+        rd_kafka_t *rk;
+        test_msgver_t *mv;
+        int64_t assigned_at;
+        int64_t revoked_at;
+        int partition_cnt;
+        int rebalance_cnt;
+        int max_rebalance_cnt;
+} _consumer_t;
+
+
+/**
+ * @brief Call poll until a rebalance has been triggered
+ */
+int static_member_wait_rebalance (_consumer_t *c, int64_t start,
+                                  int64_t *target, int timeout_ms) {
+        int64_t tmout = test_clock() + (timeout_ms * 1000);
+
+        while (timeout_ms < 0 ? 1 : test_clock() <= tmout) {
+                test_consumer_poll_once(c->rk, c->mv, 1000);
+                if (*target > start)
+                        return 1;
+              }
+
+        return 0;
+}
+
+static void rebalance_cb (rd_kafka_t *rk,
+                          rd_kafka_resp_err_t err,
+                          rd_kafka_topic_partition_list_t *parts,
+                          void *opaque) {
+        int i;
+        _consumer_t *c = opaque;
+
+        switch (err)
+        {
+        case RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS:
+                c->partition_cnt = parts->cnt;
+                c->assigned_at = test_clock();
+
+                rd_kafka_assign(rk, parts);
+
+                TEST_SAY("%s Assignment (%d partition(s)): ",
+                         rd_kafka_name(rk), parts->cnt);
+                for (i = 0 ; i < parts->cnt ; i++)
+                        TEST_SAY0("%s%s[%"PRId32"]",
+                                  i == 0 ? "" : ", ",
+                                  parts->elems[i].topic,
+                                  parts->elems[i].partition);
+                TEST_SAY0("\n");
+
+                break;
+
+        case RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS:
+                c->rebalance_cnt++;
+                TEST_ASSERT(c->rebalance_cnt == c->max_rebalance_cnt,
```

Comment:
Shouldn't this be `rebalance_cnt <= max_rebalance_cnt` ?
Otherwise max is not a max

Reply:
what about renaming it `expected_rebalance_cnt`. We want to ensure we do not rebalance too often or not often enough.

Reply:
practically it can only be set to 1 (or rebalance_cnt+1), right?
So maybe change the logic to check for which rebalance err you are getting, .e.g above the switch do:
`TEST_ASSERT(c->expected_rebalance_event == err)`

which is either ASSIGN_PART.. or REVOKE_PA..

This also makes the test more rigid since it will verify the rebalance event ordering (revoke, assign, revoke, assign)

Reply:
If I miss a rebalance that I expect to happen but it didn't I could end up with  a rebalance_cnt of 1 which is less than the expected value of 3. This would falsely advertise that static group membership is working as expected. 

We could track assign and revoke although it seems like overkill so I won't have an assignment without a revocation and I won't actually be a member of that group if I don't have an assignment.

Reply:
That's sort of my point: we don't want the test to assume and skim over important details such as the order of rebalance events, the more rigid the test the better the test coverage.

Reply:
consider it done